test/extensions/filters/http/cache/cache_filter_test.cc

```diff
@@ -114,19 +131,86 @@ class CacheFilterTest : public ::testing::Test {
     ::testing::Mock::VerifyAndClearExpectations(&decoder_callbacks_);
   }
 
+  // This function assumes that there is a body created using std::string(body_size, 'a').
+  void testSuccessfulValidation(CacheFilterSharedPtr filter, uint64_t body_size) {
+    ASSERT(body_size > 0);
+
+    // Make request require validation
+    request_headers_.setReferenceKey(Http::CustomHeaders::get().CacheControl,
+                                     Http::CustomHeaders::get().CacheControlValues.NoCache);
+
+    // Decoding the request should find a cached response that requires validation.
+    // As far as decoding the request is concerned, this is the same as a cache miss with the
+    // exception of injecting validation precondition headers.
+    testDecodeRequestMiss(filter);
+
+    // Make sure validation conditional headers are added
+    const Http::TestRequestHeaderMapImpl injected_headers = {
+        {"if-none-match", etag_}, {"if-modified-since", response_last_modified_}};
+    EXPECT_THAT(request_headers_, IsSupersetOfHeaders(injected_headers));
+
+    // Encode 304 response
+    // Advance time to make sure the cached date is updated with the 304 date
+    time_source_.advanceTimeWait(std::chrono::seconds(10));
+    const std::string not_modified_date = formatter_.now(time_source_);
+    Http::TestResponseHeaderMapImpl not_modified_response_headers = {{":status", "304"},
+                                                                     {"date", not_modified_date}};
+
+    // The filter should continue headers encoding without ending the stream as data will be
+    // injected.
+    EXPECT_EQ(filter->encodeHeaders(not_modified_response_headers, true),
+              Http::FilterHeadersStatus::ContinueAndDontEndStream);
+
+    // Check for the cached response headers with updated date
+    Http::TestResponseHeaderMapImpl updated_response_headers = response_headers_;
+    updated_response_headers.setDate(not_modified_date);
+    EXPECT_THAT(not_modified_response_headers, IsSupersetOfHeaders(updated_response_headers));
+
+    // The filter should inject data in chunks sized according to the buffer limit.
+    const int chunks_count = (body_size + buffer_limit_ - 1) / buffer_limit_;
+    // The size of all chunks except the last one is equal to the buffer_limit_.
+    EXPECT_CALL(encoder_callbacks_,
+                injectEncodedDataToFilterChain(
+                    testing::Property(&Buffer::Instance::toString,
+                                      testing::Eq(std::string(buffer_limit_, 'a'))),
```

Comment:
I was thinking about this testcase. I think it would give us more confidence in the correctness of the system if you used a buffer_limit of something much smaller, say 6. Then you can use a payload of "1234567890" and make sure the first chunk is "123456" and the second is "7890".

Right now you are comparing against a bunch of 'a's and I'm wondering if a bug could creep in where we take the wrong span of that string in the logic, but your test doesn't care.

WDYT?

Reply:
Good point! Thanks for doing a thorough review. I didn't pay much attention on the existing code but focused on downstream watermark handling test :P